allow specifying certificates used to talk to consul for storage backend

dkaffee92 opened the pull request:
resolves https://github.com/hashicorp/vault/issues/378

dkaffee92 commented:
@armon I've tried to update the dependencies and it seems I broke something, sorry for being a go noob.
How does this `godeps` lib work :-)?

ekristen commented:
+1 please add!

armon commented on physical/consul.go:
```diff
@@ -5,9 +5,12 @@ import (
 
 	"github.com/armon/go-metrics"
 	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/tlsutil"
```
We should not be pulling in this library from Consul, that is more of an internal to Consul thing. We don't need any of that functionality anyways, since we can probably just use `tls.LoadX509KeyPair`

armon commented on website/source/docs/config/index.html.md:
```diff
@@ -111,6 +111,22 @@ For Consul, the following options are supported:
 
   * `token` (optional) - An access token to use to write data to Consul.
 
+  * `tls_ca_file` (optional, mandatory when scheme is https) - The path to the CA certificate
```
These are not mandatory. If no `ca_file` is provided the system bundle (default) should be used. If `cert` and `key` are not provided, then mutual TLS is not done, but we can verify the server authenticity.

dkaffee92 replied:
@armon I'm not sure I understand what you mean by mutual TLS is not done but we can verify the server authenticity.

armon replied:
In many cases, there verification is only done in one direction. For example, you browser verifies the authenticity of the web server, but the web server does not care about the identity of the client. Similarly, you can do a one way authentication "is consul who it claims to be" without using a `cert` or `key`. Alternatively, Consul may verify the client identity which requires those. In that case the verification is done in both directions and is "mutual".

armon commented:
@dkaffee92 Left some comments! Sorry about delay, we wanted to cut 0.2 without needing to extend our QA process for new features

dkaffee92 commented:
@armon, I've refactored it,

I extracted the `TLSClientConfig` building process into a different method and also took notice to the use cases of whether you have a certificate and a private key, a ca certificate or to use the system bundle and should you want to use insecure verification.

I added some more text to the documentation to help make it a bit more clear and to link to the relevant locations in Consul's documentation.

I've built a version of this and deployed it to a test cluster of mine with consul running in `verify_incoming` and `verify_outgoing` set to true and it works as intended.

Have a look and let me know what you think.

armon commented:
Looks great, thanks!

vineet01 commented:
I am not able to get it to work .
This is what my vault configuration for consul looks like -
 
address = "10.1.1.4:8500"
path = "vault"
tls_ca_file = "/path/rootCA.pem"
advertise_addr = "https://10.1.1.4:8200/"

Is that all i need to ensure TLS connection to consul as backend ? i have verify_incoming,verify_outgoing as true in consul configuration.
If i add -> scheme = "https" to above configuration i am not able to access it anymore. That throws an error -> Get https://10.1.1.4:8500/v1/kv/vault/core/seal-config: tls: oversized record received with length 20527

jefferai commented:
@vineet01 Do you have a local Consul agent running? If so, that is the preferred way to connect to Consul.

vineet01 commented:
I was trying to configure it in a more stable way, in case local consul agent crashes/gets killed, Entire thing goes down.
Is it not possible to specify multiple backend consul instance addresses to vault? Preferably over TLS.
In case its not possible i have thought of writing a failure monitor script which will kill vault too if consul fails. So that some other vault node can take over. Does that make sense?
*scratches head

jefferai commented:
If you're using Vault in HA mode via Consul, if the local Consul agent crashes or is killed, one of the other Vault nodes should take over, because that Vault instance will fail to hold onto its leader lock.

You cannot specify multiple Consul addresses; if you are running multiple Consul instances and want to talk to them, using a load-balancer is the recommended approach. But using the local Consul agent is still recommended overall.

vineet01 commented:
It happens but takes upto 2-3 Minutes (i am running 3 Vault instances, 2 Servers, 3 Agents). I keep getting Leader <none> for upto 2-3 Minutes.

High-Availability Enabled: true                                        
Mode: standby
Leader: <none> 

Is there a way to minimize this delay or make it instantaneous?

jefferai commented:
That's probably because you have 2 Consul servers. Consul needs an odd number of servers in order to have a quorum, so with 2 you may be getting behavior where the nodes aren't sure which of them should be leader.

On our cluster the transition happens in < 1 second. It's a 3-node Consul cluster and a 3-node Vault cluster.

vineet01 commented:
Ah ok. Will try that. Thanks :)
